pkg/pod/pods.go

```diff
@@ -1075,76 +870,304 @@ func (p *pod) getDirNames(path string) ([]string, error) {
+		return p.preparedPath()
+	} else if p.isExitedGarbage {
+		return p.exitedGarbagePath()
+	} else if p.isGarbage {
+		return p.garbagePath()
+	} else if p.isGone {
+		return "" // TODO(vc): anything better?
+	}
+
+	return p.runPath()
+}
+
+// Stage1RootfsPath returns the stage1 path of the pod.
+func (p *Pod) Stage1RootfsPath() (string, error) {
+	stage1RootfsPath := "stage1/rootfs"
+	if p.UsesOverlay() {
+		stage1TreeStoreID, err := p.GetStage1TreeStoreID()
 		if err != nil {
-			stderr.PrintE(fmt.Sprintf("unable to get status of app %q", name), err)
-			continue
+			return "", err
 		}
-		stats[name] = s
+		stage1RootfsPath = fmt.Sprintf("overlay/%s/upper/", stage1TreeStoreID)
 	}
-	return stats, nil
+
+	return filepath.Join(p.Path(), stage1RootfsPath), nil
 }
 
-// sync syncs the pod data. By now it calls a syncfs on the filesystem
-// containing the pod's directory.
-func (p *pod) sync() error {
-	cfd, err := p.Fd()
+// JournalLogPath returns the path to the journal log dir of the pod.
+func (p *Pod) JournalLogPath() (string, error) {
+	stage1RootfsPath, err := p.Stage1RootfsPath()
 	if err != nil {
-		return errwrap.Wrap(fmt.Errorf("error acquiring pod %v dir fd", p.uuid.String()), err)
-	}
-	if err := sys.Syncfs(cfd); err != nil {
-		return errwrap.Wrap(fmt.Errorf("failed to sync pod %v data", p.uuid.String()), err)
+		return "", err
 	}
-	return nil
+	return filepath.Join(stage1RootfsPath, "/var/log/journal/"), nil
 }
 
-// overlayPrepared returns true if the given pod was prepared using overlay else false.
-// It returns an error if the operation failed.
-func (p *pod) overlayPrepared() (bool, error) {
-	_, err := os.Stat(filepath.Join(p.path(), common.OverlayPreparedFilename))
-	if os.IsNotExist(err) {
-		return false, nil
+// State returns the current state of the pod
+func (p *Pod) State() string {
+	state := Running
+
+	if p.isEmbryo {
+		state = Embryo
+	} else if p.isPreparing {
+		state = Preparing
+	} else if p.isAbortedPrepare {
+		state = AbortedPrepare
+	} else if p.isPrepared {
+		state = Prepared
+	} else if p.isDeleting {
+		state = Deleting
+	} else if p.isExitedDeleting {
+		state = ExitedDeleting
+	} else if p.isExited { // this covers p.isExitedGarbage
+		state = Exited
+		if p.isExitedGarbage {
+			state = ExitedGarbage
+		}
+	} else if p.isGarbage {
+		state = Garbage
 	}
+
+	return state
+}
+
+// isRunning does the annoying tests to infer if a pod is in a running state
+func (p *Pod) isRunning() bool {
+	// when none of these things, running!
+	return !p.isEmbryo && !p.isAbortedPrepare && !p.isPreparing && !p.isPrepared &&
+		!p.isExited && !p.isExitedGarbage && !p.isExitedDeleting && !p.isGarbage && !p.isDeleting && !p.isGone
+}
+
+// AfterRun tests if a pod is in a post-running state
+func (p *Pod) AfterRun() bool {
+	return p.isExitedDeleting || p.isDeleting || p.isExited || p.isGarbage
+}
+
+// GetAppExitCode returns the app's exit code.
+// Returns error if the exit code file doesn't exit.
```

Comment:
Is this really the only error case? I see that `readIntFromFile` also returns an error if parsing failed.